PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+            return
+        }
+
+        self.products(withIdentifiers: [productIdentifier]) { products in
+            self.postReceipt(withTransaction: transaction,
+                             receiptData: receiptData,
+                             products: Set(products))
+        }
+    }
+
+    func postReceipt(withTransaction transaction: SKPaymentTransaction, receiptData: Data, products: Set<SKProduct>) {
+        var maybeProductInfo: ProductInfo?
+        var maybePresentedOfferingID: String?
+        if let product = products.first {
+            let productInfo = ProductInfoExtractor().extractInfo(from: product)
+            let productID = productInfo.productIdentifier
+            let presentedOfferingID = presentedOfferingIDsByProductID[productID]
+            presentedOfferingIDsByProductID.removeValue(forKey: productID)
+            maybeProductInfo = productInfo
+            maybePresentedOfferingID = presentedOfferingID
+        }
+        let unsyncedAttributes = unsyncedAttributes
+
+        backend.post(receiptData: receiptData,
+                     appUserID: appUserID,
+                     isRestore: allowSharingAppStoreAccount,
+                     productInfo: maybeProductInfo,
+                     presentedOfferingIdentifier: maybePresentedOfferingID,
+                     observerMode: !finishTransactions,
+                     subscriberAttributes: unsyncedAttributes) { maybePurchaserInfo, maybeError in
+            self.handleReceiptPost(withTransaction: transaction,
+                                   maybePurchaserInfo: maybePurchaserInfo,
+                                   maybeSubscriberAttributes: unsyncedAttributes,
+                                   maybeError: maybeError)
+        }
+    }
+
+    func handleReceiptPost(withTransaction transaction: SKPaymentTransaction,
+                           maybePurchaserInfo: PurchaserInfo?,
+                           maybeSubscriberAttributes: SubscriberAttributeDict?,
+                           maybeError: Error?) {
+        operationDispatcher.dispatchOnMainThread {
+            let appUserID = self.appUserID
+            self.markSyncedIfNeeded(subscriberAttributes: maybeSubscriberAttributes,
+                                    appUserID: appUserID,
+                                    maybeError: maybeError)
+
+            let maybeCompletion = self.getAndRemovePurchaseCompletedCallback(forTransaction: transaction)
+            let nsError = maybeError as NSError?
+            let finishable = (nsError?.userInfo[ErrorDetails.finishableKey as String] as? NSNumber)?.boolValue ?? false
+            if let purchaserInfo = maybePurchaserInfo {
+                self.purchaserInfoManager.cache(purchaserInfo: purchaserInfo, appUserID: appUserID)
+                maybeCompletion?(transaction, purchaserInfo, nil, false)
+
+                if self.finishTransactions {
+                    self.storeKitWrapper.finishTransaction(transaction)
+                }
```

Comment:
could pull that block into a defer?

Reply:
There's one single return from that func that shouldn't run that code 😢 
```swift
} else if !finishable {
                Logger.error(Strings.receipt.unknown_backend_error)
                maybeCompletion?(transaction, nil, maybeError, false)
}
```            
(Which really should be just an `else` instead of `else if` but I'm trying to remain true to the old ObjC 🙃 )

Reply:
Oh Wait, maybe you meant I could just check for `finishable` in the defer.... yeahhhhhh

Reply:
Ooof, this one is worse than it seems, here's the logic we'd need:
```swift
            defer {
                if (finishable && self.finishTransactions) || (maybePurchaserInfo != nil && self.finishTransactions) {
                    self.storeKitWrapper.finishTransaction(transaction)
                }
            }
```

I'll leave it as is since this is more confusing IMO.

Reply:
maybe i got a little over-zealous about defer 😅